pkg/ruby/check_upgrade.go

```diff
@@ -0,0 +1,199 @@
+		}
+	}
+	return "", fmt.Errorf("could not find gemspec file in repo")
+}
+
+// cachedGemspecPath returns the path to a cached gemspec file
+func (o *Options) cachedGemspecPath(pkg *Package, gemspec string) string {
+	return path.Join(rubyCacheDirectory, pkg.Name, fmt.Sprintf("%s-%s", pkg.Ref, gemspec))
+}
+
+// fetchFile downloads a given file (in this case a gemspec) from a given
+// Github repository. It downloads the file from the branch or tag specified
+// in the melange yaml. It will cache the file using the name of the file and
+// the branch or tag specified in the melange yaml.
+func (o *Options) fetchFile(pkg *Package, file string) (string, error) {
+	logger := log.New(log.Writer(), "wolfictl ruby check-upgrade: ", log.LstdFlags|log.Lmsgprefix)
+	cachedPath := o.cachedGemspecPath(pkg, file)
+	cached, err := os.Open(cachedPath)
+	if err != nil || o.NoCache {
+		ctx := context.Background()
+
+		client := github.NewClient(o.Client.Client)
+		gitOpts := gh.GitOptions{
+			GithubClient: client,
+			Logger:       logger,
+		}
+
+		fileContent, err := gitOpts.RepositoryFilesContents(ctx, pkg.Repo.Organisation, pkg.Repo.Name, file, pkg.Ref)
+		if err != nil {
+			return "", err
+		}
+
+		// Decode the base64-encoded content
+		decodedContent, err := base64.StdEncoding.DecodeString(*fileContent.Content)
+		if err != nil {
+			return "", fmt.Errorf("error decoding file content: %w", err)
+		}
+
+		err = os.MkdirAll(path.Dir(cachedPath), 0o755)
+		if err != nil {
+			return "", fmt.Errorf("failed to create cache directory: %w", err)
+		}
+
+		cached, err = os.Create(cachedPath)
+		if err != nil {
+			return "", fmt.Errorf("failed to create cache file: %w", err)
+		}
+
+		_, err = cached.Write(decodedContent)
+		if err != nil {
+			return "", fmt.Errorf("failed to write cache file: %w", err)
+		}
+	}
+	defer cached.Close()
+	return cachedPath, nil
+}
+
+// checkVersionConstraint takes a gemspec and extracts the version constraints
+// and checks them against the RubyUpdateVersion. This function attempts to use
+// inline ruby to determine if the version is within constraints. If ruby is
+// not installed on the system then the terraform go-version package is used as
+// it has very similar version constraints. The inline ruby is the preferred
+// method in case something changes (highly unlikely).
+//
+// NOTE: If there is no required_ruby_version found in the gemspec then we
+//
+//	assume there is no limit.
+func (o *Options) checkVersionConstraint(gemspecFile string) error {
+	cached, err := os.Open(gemspecFile)
+	if err != nil {
+		return fmt.Errorf("opening gemspec file: %w", err)
+	}
+	versionLine := ""
+	scanner := bufio.NewScanner(cached)
+	for scanner.Scan() {
+		if strings.Contains(scanner.Text(), requiredRubyVersionKey) {
+			versionLine = scanner.Text()
+			break
+		}
+	}
+
+	// Could not find required_ruby_version in gemspec, assuming no constraint
+	if versionLine == "" {
+		return nil
+	}
+
+	pattern := `["']([^"']+)["']`
```

Comment:
A few questions here:
* is the sole purpose of this regex to pull the version out of a constraint?
* is there any risk that the version string could be malformed (i.e. `"1.6.0'`)?
    * if malformed, do you still want to pull the version?

Reply:
Good questions! It is, I'd expect the version constraint to be properly formed since it's coming from the .gemspec which is valid ruby and would fail the gem build. Even if it was malformed I'd prefer to bias towards extracting as many version constraints as we can

Reply:
Ah right, that makes sense. I forgot that would have to pass the scrutiny of build.